src/java.net.http/share/classes/jdk/internal/net/http/LimitingSubscriber.java

```diff
@@ -0,0 +1,148 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.net.http;
+
+import jdk.internal.net.http.ResponseSubscribers.TrustedSubscriber;
+import jdk.internal.net.http.common.Utils;
+
+import java.io.IOException;
+import java.net.http.HttpResponse.BodySubscriber;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow.Subscription;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A subscriber limiting the maximum number of bytes that are allowed to be consumed by a downstream subscriber.
+ *
+ * @param <T> the response type
+ */
+public final class LimitingSubscriber<T> implements TrustedSubscriber<T> {
+
+    private final BodySubscriber<T> downstreamSubscriber;
+
+    private final long capacity;
+
+    private State state;
+
+    private long length;
+
+    private interface State {
+
+        enum Terminated implements State { INSTANCE }
+
+        record Subscribed(Subscription subscription) implements State {}
+
+    }
+
+    /**
+     * @param downstreamSubscriber the downstream subscriber to pass received data to
+     * @param capacity the maximum number of bytes that are allowed
+     * @throws IllegalArgumentException if {@code capacity < 0}
+     */
+    public LimitingSubscriber(BodySubscriber<T> downstreamSubscriber, long capacity) {
+        if (capacity < 0) {
+            throw new IllegalArgumentException("was expecting \"capacity >= 0\", found: " + capacity);
+        }
+        this.downstreamSubscriber = requireNonNull(downstreamSubscriber, "downstreamSubscriber");
+        this.capacity = capacity;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (state != null) {
+            subscription.cancel();
+        } else {
+            state = new State.Subscribed(subscription);
+            downstreamSubscriber.onSubscribe(subscription);
+        }
+    }
+
+    @Override
+    public void onNext(List<ByteBuffer> buffers) {
+
+        // Check arguments
+        requireNonNull(buffers, "buffers");
+        assert Utils.hasRemaining(buffers);
+
+        // Short-circuit if not subscribed
+        if (!(state instanceof State.Subscribed subscribed)) {
+            return;
+        }
+
+        // See if we may consume the input
+        boolean lengthAllocated = allocateLength(buffers);
+        if (lengthAllocated) {
+            downstreamSubscriber.onNext(buffers);
+        }
+
+        // Otherwise, trigger failure
+        else {
```

Comment:
Nit - this `else` feels a bit "far" away from the corresponding `if`. Perhaps you could consider changing it to something like:
```java
if {
    ...
} else { // Otherwise, trigger failure
    ...
}
```
or even,

```java
if {
    ...
} else { 
    // Otherwise, trigger failure
    ...
}
```

Reply:
Replaced as suggested in 23f79a20476f784058801454a28d37f37df92982.

IMHO, the suggested style causes misalignment on the comments associated with `if` and `else`. That is, the comments of the `else` block gets indented deeper, even though it is semantically at the same level with the `if` comment:

![image](https://github.com/user-attachments/assets/da465a30-284a-40f2-948a-fa5e1f79c53d)